File: lib/log.mjs
/**
 * @fileoverview log.mjs — structured logging shared by this repo's node tools.
 *
 * One JSON object per line on stdout (LOG_FORMAT=logfmt for logfmt):
 *   {"ts":"2025-09-02T12:00:00Z","level":"info","tool":"generate-pi-coding-agent","msg":"...","key":"value"}
 *
 * No level filtering happens here, ever — debug/trace/warn/error all go to the
 * stream; a consumer that wants less noise filters downstream with jsonlines
 * tooling (`jq 'select(.level=="error")'`). The producer's only judgment call
 * is WHICH level to label a line, never whether to drop it (docs/d045).
 *
 * Stream: stdout by default. A script whose stdout IS the machine-consumed
 * payload (git/audit.mjs's JSON report, search-references' results,
 * -vbs-mirror-all.sh's --list TSV) calls setLogStream("stderr") once at
 * startup so logs cannot interleave with the payload; `2>&1 | jq` still gives
 * the consumer one jsonlines stream. Docs/d045 owns the policy.
 *
 * Error values in fields serialize as structured objects — name, message,
 * stack, own enumerable properties — with the full cause/suppressed chain
 * walked recursively (`cause`, AggregateError `errors`, SuppressedError
 * `error`/`suppressed`). Messages are never interpolated with dynamic values:
 * the message is the stable label, the fields are the metadata (docs/d045).
 *
 * Env:
 *   LOG_FORMAT  json|logfmt   (default: json)
 *   LOG_TOOL    component name (default derived per call site; each tool sets
 *               it once: setLogTool("coding-agent/generate"))
 *
 * Import via the LIB_DIR convention (docs/d023) so the same file works both in
 * place and when staged next to the generators:
 *   const { logInfo, logWarn } = await import(`${LIB_DIR}/log.mjs`);
 */

let tool = process.env.LOG_TOOL ?? "node";
/**
 * @param {string} name
 */
export function setLogTool(name) {
	tool = name;
}

let stream = /** @type {{ write(chunk: string): unknown }} */ (
	/** @type {unknown} */ (process.stdout)
);
/**
 * Route the log stream away from stdout for scripts whose stdout is a
 * machine-consumed payload (docs/d045). "stdout" restores the default.
 * @param {"stdout"|"stderr"} name
 */
export function setLogStream(name) {
	stream = name === "stderr" ? process.stderr : process.stdout;
}

function ts() {
	return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

// Cause/suppressed chains are user data; a cyclic one must not hang the emit.
// No real chain here is deeper than a wrapped fetch → wrapped parse → cause.
const MAX_SERIALIZATION_DEPTH = 8;

/**
 * Serialize a log field value. Errors become structured objects (name,
 * message, stack, own enumerable properties, plus the explicitly walked
 * non-enumerable `cause`/`errors`/`error`/`suppressed`); plain objects and
 * arrays recurse; primitives pass through.
 * @param {unknown} v
 * @param {number} depth
 * @returns {unknown}
 */
function serialize(v, depth) {
	if (depth >= MAX_SERIALIZATION_DEPTH) {
		return v instanceof Error
			? { name: v.name, message: v.message }
			: Array.isArray(v)
				? "[max depth]"
				: { "[max depth]": true };
	}
	if (v instanceof Error) return serializeError(v, depth);
	if (Array.isArray(v)) return v.map((item) => serialize(item, depth + 1));
	if (v !== null && typeof v === "object") {
		return Object.fromEntries(
			Object.entries(v).map(([k, item]) => [k, serialize(item, depth + 1)]),
		);
	}
	return v;
}

/**
 * `cause`, AggregateError's `errors`, and SuppressedError's `error`/`suppressed`
 * are all NON-enumerable own properties (verified on node 24) — Object.entries
 * misses them, so they are walked explicitly. Attached properties from
 * `Object.assign(err, {...})` ARE enumerable and flow through the general pass.
 * @param {Error} err
 * @param {number} depth
 * @returns {unknown}
 */
function serializeError(err, depth) {
	/** @type {Record<string, unknown>} */
	const out = { name: err.name, message: err.message };
	if (typeof err.stack === "string") out.stack = err.stack;
	for (const [k, v] of Object.entries(err)) {
		out[k] = serialize(v, depth + 1);
	}
	for (const k of ["cause", "errors", "error", "suppressed"]) {
		if (k in out || !(k in err)) continue;
		const v = /** @type {Record<string, unknown>} */ (
			/** @type {unknown} */ (err)
		)[k];
		out[k] = serialize(v, depth + 1);
	}
	return out;
}

/**
 * @param {unknown} v
 */
function fmtLogfmtValue(v) {
	// Structured fields may be objects/arrays/errors after serialization —
	// JSON-encode those instead of leaking "[object Object]".
	if (v !== null && typeof v === "object") return JSON.stringify(v);
	const s = String(v);
	return /[\s"=\\]/.test(s) || s === "" ? JSON.stringify(s) : s;
}

/**
 * @param {string} level
 * @param {string} msg
 * @param {Record<string, unknown>} [fields]
 */
function emit(level, msg, fields = {}) {
	const base = { ts: ts(), level, tool, msg };
	if (process.env.LOG_FORMAT === "logfmt") {
		const rest = Object.entries(
			/** @type {Record<string, unknown>} */ (serialize(fields, 0)),
		)
			.map(([k, v]) => `${k}=${fmtLogfmtValue(v)}`)
			.join(" ");
		const msgv = /[\s"=\\]/.test(msg) ? JSON.stringify(msg) : msg;
		stream.write(
			`ts=${base.ts} level=${level} tool=${tool} msg=${msgv}${rest ? " " + rest : ""}\n`,
		);
		return;
	}
	stream.write(
		JSON.stringify({
			...base,
			.../** @type {Record<string, unknown>} */ (serialize(fields, 0)),
		}) + "\n",
	);
}

/**
 * @param {string} msg
 * @param {Record<string, unknown>} [fields]
 */
export const logDebug = (msg, fields) => emit("debug", msg, fields);
/**
 * @param {string} msg
 * @param {Record<string, unknown>} [fields]
 */
export const logInfo = (msg, fields) => emit("info", msg, fields);
/**
 * @param {string} msg
 * @param {Record<string, unknown>} [fields]
 */
export const logWarn = (msg, fields) => emit("warn", msg, fields);
/**
 * @param {string} msg
 * @param {Record<string, unknown>} [fields]
 */
export const logError = (msg, fields) => emit("error", msg, fields);

File: lib/log.py
"""Structured logging shared by this repo's python tools — same contract as
lib/log.sh / lib/log.mjs: one JSON object per line on stdout.

    {"ts":"2025-09-02T12:00:00Z","level":"info","tool":"local-llm/upkeep","msg":"...","key":"value"}

Env:
    LOG_FORMAT  json|logfmt   (default: json)
    LOG_TOOL    component name (default "python"; set per tool with set_tool(),
                e.g. log.set_tool("local-llm/upkeep"))

No level filtering happens here, ever — debug/trace/warn/error all go to the
stream; a consumer filters downstream with jsonlines tooling (docs/d045).
Stream is stdout by default; a script whose stdout IS the machine-consumed
payload calls set_stream(sys.stderr) once at startup so logs cannot interleave
with the payload (docs/d045).

Usage from a PEP-723 `uv run` script (the script's dir is on sys.path, the
repo root is not — bootstrap by inserting ../lib):

    import pathlib, sys
    sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent / "lib"))
    import log
    log.set_tool("local-llm/upkeep")
    log.info("pulled repo", repo="org/name", revision="abc123")
"""
import datetime
import json
import os
import re
import sys
import traceback

_tool = os.environ.get("LOG_TOOL", "python")
_stream = sys.stderr if os.environ.get("LOG_STREAM") == "stderr" else sys.stdout

# Same policy as lib/log.mjs (docs/d045): exceptions in fields serialize as
# structured objects — name, message, traceback, and the explicit __cause__
# chain — never as interpolated prose. python's implicit __context__ is left
# out: it fires for ANY exception raised inside an except block, which would
# attach unrelated noise to every logged failure.
def _serialize(obj, depth=0):
    if isinstance(obj, BaseException):
        out = {"name": type(obj).__name__, "message": str(obj)}
        tb = "".join(traceback.format_exception(type(obj), obj, obj.__traceback__))
        if tb:
            out["traceback"] = tb
        if depth < 8 and obj.__cause__ is not None:
            out["cause"] = _serialize(obj.__cause__, depth + 1)
        return out
    return str(obj)


def set_tool(name: str) -> None:
    global _tool
    _tool = name


def set_stream(stream) -> None:
    """Route logs away from stdout when this script's stdout is a
    machine-consumed payload (docs/d045); pass sys.stderr."""
    global _stream
    _stream = stream


def emit(level: str, msg: str, **fields) -> None:
    ts = datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    if os.environ.get("LOG_FORMAT") == "logfmt":
        def val(v) -> str:
            s = str(v)
            return json.dumps(s) if re.search(r'[\s"=\\]', s) or s == "" else s
        rest = " ".join(f"{k}={val(v)}" for k, v in fields.items())
        m = json.dumps(msg) if re.search(r'[\s"=\\]', msg) else msg
        line = f"ts={ts} level={level} tool={_tool} msg={m}"
        if rest:
            line += " " + rest
    else:
        rec = {"ts": ts, "level": level, "tool": _tool, "msg": msg, **fields}
        line = json.dumps(rec, ensure_ascii=False, default=lambda o: _serialize(o))
    print(line, file=_stream)


def debug(msg: str, **fields) -> None:
    emit("debug", msg, **fields)


def info(msg: str, **fields) -> None:
    emit("info", msg, **fields)


def warn(msg: str, **fields) -> None:
    emit("warn", msg, **fields)


def error(msg: str, **fields) -> None:
    emit("error", msg, **fields)
